core/src/credential.rs

```diff
@@ -0,0 +1,147 @@
+
+#![allow(unused_variables)]
+#![allow(dead_code)]
+use std::{error::Error, time::{Instant, SystemTime}};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+// cred_subject is a generic that implements trait X
+// trait X allows us to encode that object into JSON-LD
+// We provide types that implement trait X for the cred types that we support
+// Users can also user their own types that implement trait X if they need a different structure
+// --- 
+// Default context and Cred types are defaulted but can be redefined 
+
+type VCContext = [&'static str];
+
+pub(crate) const JSON_LD_CONTEXT:  VCContext = ["https://www.w3.org/2018/credentials/v1", "https://www.w3.org/2018/credentials/examples/v1"];
+
+pub const CRED_TYPE_PERMANENT_RESIDENT_CARD: str = "PermanentResidentCard";
+pub const CRED_TYPE_BANK_CARD: str = "BankCard";
+
+type CredentialSubject = serde_json::Value;
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct VerifiableCredential <'a> {
+    #[serde(rename = "@context")]
+	context:  VCContext,
+    #[serde(rename = "@id")]
+	id: String,
+    cred_type:  String,
+	issuer: String,
+	issuance_date: SystemTime,
+	subject: CredentialSubject,
+}
+
+pub struct IssuedCredential <'a> {
+    verifiable_credential:VerifiableCredential,
+    proof: *mut CredentialProof,
+}
+
+pub struct CredentialManager {
+    credential_subject: HashMap<&str, &str>,
```

Comment:
Shouldn't the return value here be `json_serde::Value`?

Reply:
The idea is to maintain string in template and convert to `json_serde::Value` later, maybe we need to convert it to interim state to change some fields.

Reply:
As far as i remember there was a discussion with @fkim7 where we were discussing an opportunity to have some kind of credential_subject templates dependent on the credential types, correct me if I'm wrong. The implementation allows end-users to redefine that via trait but by default, we provide two credential types and two corresponding credential subjects.

Reply:
Here is a link for the reference (see above also) https://github.com/knox-networks/ssi/pull/5#discussion_r908461052

Reply:
I think we had decided against it. The end user might use templates but we should expect the full credential subject to be passed down to the `create_credential` function. I think the templating approach is too complex/cumbersome at this level of abstraction